### PR TITLE
Dataset remix

### PR DESCRIPTION
**Pull request by dirkgr**
This depends on #5344 being merged first.

It adds a step that can remix an existing dataset in a flexible way. You give it a `Dict[str, str]` like this:
```
{
    "train": "train + dev[:10000]",
    "validation": "dev[10000:]"
}
```
That means, from the original dataset, take the whole "train" split, combine it with the first 10000 instances of the "dev" split, and call that the new "train" split. Also, take everything after the first 10000 instances of the "dev" split, and call that the "validation" split.

Also, there are some urgent Tango fixes in here.

### CODE REVIEW

File: allennlp/tango/dataset.py [hunk 1]
```diff
@@ -72,3 +74,46 @@ def run(self, reader: DatasetReader, splits: Dict[str, str]) -> DatasetDict:  #
...
+
+    DETERMINISTIC = True
+    CACHEABLE = False  # This is so fast it's not worth caching.
+    VERSION = "001"
+
+    def run(  # type: ignore
+        self, input: DatasetDict, new_splits: Dict[str, str], keep_old_splits: bool = True
+    ) -> DatasetDict:
+        def get_slice(split_name: str) -> Sequence[Any]:
+            slice_match = re.match(r"(.*)\[([0123456789:]*)]", split_name)
```
Comment (AkshitaB) [inline allennlp/tango/dataset.py:93]: This won't work for something like `train[:50000] + dev[:10000]`. Is it supposed to?

Comment (dirkgr) [reply allennlp/tango/dataset.py:93]: Yes, it will work. This function is only called on the parts after `.split("+")`.

Comment (AkshitaB) [reply allennlp/tango/dataset.py:93]: Ok I see, and train[:50000] in that case is interpreted as it should be?

Comment (dirkgr) [reply allennlp/tango/dataset.py:93]: Yes. It supports full Python slice syntax on line 98. In condensed form, it does `slice(*match.split(":"))`.

**Review by AkshitaB — APPROVED** [review]

**Pull request merged**